Added test for iommu

dhsrivas opened the pull request:
Various combination of unbind, bind, change iommu group domain type, reset and rescan is used to form sub-tests that test and exercise iommu code.

Signed-off-by: Dheeraj Kumar Srivastava <dheerajkumar.srivastava@amd.com>

dhsrivas commented:
**_Test result sample_**
#sudo avocado run iommu_tests.py -p pci_devices="0000:03:00.3" -p count=3 --test-runner runner
JOB ID     : 7304e36afa5aae5d04a22dfaf198a43f1c07681c
JOB LOG    : /root/avocado/job-results/job-2022-10-04T11.18-7304e36/job.log
 (1/6) iommu_tests.py:IommuTest.test_unbind_bind: PASS (10.11 s)
 (2/6) iommu_tests.py:IommuTest.test_unbind_changedomain_bind: PASS (196.00 s)
 (3/6) iommu_tests.py:IommuTest.test_unbind_changedomain_ntimes_bind: PASS (587.90 s)
 (4/6) iommu_tests.py:IommuTest.test_unbind_bind_rescan: PASS (15.14 s)
 (5/6) iommu_tests.py:IommuTest.test_unbind_changedomain_bind_rescan: PASS (225.99 s)
 (6/6) iommu_tests.py:IommuTest.test_reset_rescan: PASS (10.13 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 1045.63 s

[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/9706483/job.log)

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+    def unbind(self, driver, pci_addr):
+        """
+        unbind the device
+        """
```
@dhsrivas please add the docstring using the PEP-257 style

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+    def change_domain(self, dom, defdom, pci_addr):
+        """
+        change the domain of device to dom
+        """
```
same here

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        """
+        change the domain of device to dom
+        """
+        cmd = f'echo {dom} > /sys/bus/pci/devices/{pci_addr}/iommu_group/type'
```
I think you can use genio.write_file() method here.

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        cmd = f'echo {dom} > /sys/bus/pci/devices/{pci_addr}/iommu_group/type'
+        process.run(cmd, shell=True, sudo=True)
+        time.sleep(5)
+        cmd = f"cat /sys/bus/pci/devices/{pci_addr}/iommu_group/type"
```
similarly genio.read() here.

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        for line in output.splitlines():
+            out = line.rsplit(None, 1)[-1]
+        if dom != "auto":
+            if out != dom:
```
can we merge this with `and` operator?

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+                self.log.info("successfully changed iommu group "
+                              "domain to {dom}")
+        else:
+            if out != defdom:
```
similarly `elif` here

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+    def bind(self, driver, pci_addr):
+        """
+        bind the device to driver
+        """
```
same for docstring as mentioned above

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        cmd = f"echo -n {pci_addr} > /sys/bus/pci/drivers/{driver}/bind"
+        process.run(cmd, shell=True, sudo=True)
+        time.sleep(5)
+        cmd = f'ls /sys/bus/pci/drivers/{driver}'
```
please use os.listdir

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+    def reset(self, pci_addr):
+        """
+        remove the device
+        """
```
same here for docstring

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        """
+        remove the device
+        """
+        cmd = f"echo '1' > /sys/bus/pci/devices/{pci_addr}/remove"
```
use genio.write()

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        """
+        rescan the system
+        """
+        cmd = "echo '1' > /sys/bus/pci/rescan"
```
genio.write_file() can be used instead.

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        """
+        get device parameter
+        driver, group, default domain(defdom)
+        """
```
same here for docstring

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        driver = pci.get_driver(pci_addr)
+
+        defdom = None
+        cmd = f"cat /sys/bus/pci/drivers/{driver}/{pci_addr}/iommu_group/type"
```
genio.read()

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+    def check(self, defdom, pci_addr):
+        """
+        check if the PCI device is in default domain
+        """
```
docstring

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+        """
+        check if the PCI device is in default domain
+        """
+        cmd = f"cat /sys/bus/pci/devices/{pci_addr}/iommu_group/type"
```
genio.read()

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,367 @@
+            # binding the driver
+            self.bind(driver, pci_addr)
+            if self.return_code == 1:
+                self.fail(f'{pci_addr} not unbounded')
```
I think this check has to be done after `self.unbind(driver, pci_addr)`

Naresh-ibm commented on io/iommu/iommu_tests.py.data/README.txt:
```diff
@@ -0,0 +1,7 @@
+
+Inputs Needed (in multiplexer file):
+------------------------------------
+pci_devices -      can be fetched from <lspci -nnD>  output. use space( ) for multiple devices "001b:62:00.0 001b:62:00.1"
```
multiple pci_addresses instead of device

Naresh-ibm commented on io/iommu/iommu_tests.py.data/README.txt:
```diff
@@ -0,0 +1,7 @@
+Inputs Needed (in multiplexer file):
+------------------------------------
+pci_devices -      can be fetched from <lspci -nnD>  output. use space( ) for multiple devices "001b:62:00.0 001b:62:00.1"
+count -      This is an interger value give for number of time tests to run.
```
integer / interger,   given / give

dhsrivas commented:
Hi Naresh,
I have address all your comments with other enhancements.

Output:
#sudo avocado run iommu_tests.py -p pci_devices="0000:03:00.3" -p count=3 --test-runner runner
[sudo] password for amd:
JOB ID     : 50922fa5123a029863a083e32313c67e384f42fe
JOB LOG    : /root/avocado/job-results/job-2022-10-25T05.49-50922fa/job.log
 (1/6) iommu_tests.py:IommuTest.test_unbind_bind: PASS (10.12 s)
 (2/6) iommu_tests.py:IommuTest.test_unbind_changedomain_bind: PASS (195.55 s)
 (3/6) iommu_tests.py:IommuTest.test_unbind_changedomain_ntimes_bind: PASS (586.49 s)
 (4/6) iommu_tests.py:IommuTest.test_unbind_bind_rescan: PASS (15.14 s)
 (5/6) iommu_tests.py:IommuTest.test_unbind_changedomain_bind_rescan: PASS (225.70 s)
 (6/6) iommu_tests.py:IommuTest.test_reset_rescan: PASS (10.13 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 1043.46 s

[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/9857584/job.log)

dhsrivas commented:
Corrected some log related issues.

Latest run log file : 
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/9859148/job.log)

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,278 @@
+        """Bind the device to driver"""
+        genio.write_file(f"/sys/bus/pci/drivers/{driver}/bind", pci_addr)
+        time.sleep(5)
+        os.listdir(f'/sys/bus/pci/drivers/{driver}')
```
same here.

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,278 @@
+            self.log.info("successfully changed iommu group domain to %s", dom)
+
+    def bind(self, driver, pci_addr):
+        """Bind the device to driver"""
```
same here for docstring

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,278 @@
+            self.log.info("successfully binded back %s", pci_addr)
+
+    def reset(self, pci_addr):
+        """Remove the device"""
```
same docstring here

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,278 @@
+            self.log.info("successfully removed the device %s", pci_addr)
+
+    def rescan(self, pci_addr):
+        """Rescan the system"""
```
same here, docstring

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,278 @@
+            self.log.info("successfully rescaned for the device %s", pci_addr)
+
+    def get_params(self, pci_addr):
+        """Get device parameter-driver, group, default domain(def_dom)"""
```
docstring

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,278 @@
+        return driver, def_dom
+
+    def check(self, def_dom, pci_addr, driver):
+        """Check if the PCI device is in default domain"""
```
same here

dhsrivas commented:
Hi Naresh

Thanks for review.

On 10/27/2022 10:53 AM, Naresh wrote:
> ***@***.**** requested changes on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In io/iommu/iommu_tests.py 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23discussion_r1006409629&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152749389%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=%2Bzmd4IpsunifSANqh5x0t9TJwz1EpdEIqQ%2BWLbzwIII%3D&reserved=0>:
> 
>> +                            "domain type change for such device")
> +        cmd = "sudo dmesg -C"
> +        process.run(cmd, ignore_status=True, shell=True)
> +
> +    def unbind(self, driver, pci_addr):
> +        """Unbind the device"""
> +        genio.write_file(f'/sys/bus/pci/drivers/{driver}/unbind', pci_addr)
> +        time.sleep(5)
> +        os.listdir(f'/sys/bus/pci/drivers/{driver}')
> +        if os.path.exists(f"/sys/bus/pci/drivers/{driver}/{pci_addr}"):
> +            self.fail(f'{pci_addr} not unbounded')
> +        else:
> +            self.log.info("successfully unbinded %s", pci_addr)
> +
> +    def change_domain(self, dom, def_dom, pci_addr):
> +        """Change the domain of device to dom"""
> 
> @dhsrivas 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdhsrivas&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152749389%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=rJFEusdvvF7Vt8Gn9EFZZn%2BzY0TmvDK%2B9WSL7w%2BIHK0%3D&reserved=0> please have the dockstring accorind to Pep-257 standard.
> 

Can you please explain me what you are referring by Pep-257 standard for 
docstring.---------------------------------------------------------------------
> 
> In io/iommu/iommu_tests.py 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23discussion_r1006409835&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152749389%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=uv1xesBAo31nj%2FhOaCzewpf3gEp%2BwWM%2FGEppdeYS1Xs%3D&reserved=0>:
> 
>> +            if driver is None:
> +                self.cancel("Device passed is not bound to any driver")
> +            lst = os.listdir(f'/sys/bus/pci/drivers/{driver}/{pci_addr}/'
> +                             'iommu_group/devices/')
> +            if len(lst) != 1:
> +                self.cancel(f"{pci_addr} belongs to iommu group having more "
> +                            "than one device but system doesnot support "
> +                            "domain type change for such device")
> +        cmd = "sudo dmesg -C"
> +        process.run(cmd, ignore_status=True, shell=True)
> +
> +    def unbind(self, driver, pci_addr):
> +        """Unbind the device"""
> +        genio.write_file(f'/sys/bus/pci/drivers/{driver}/unbind', pci_addr)
> +        time.sleep(5)
> +        os.listdir(f'/sys/bus/pci/drivers/{driver}')
> 
> I think this line is not required I feel. re-check on it, as your 
> checking the same in below line
> 

Added for logging purpose. In addition to failing based on check (Below 
lines), by this line we are logging what are devices binded to driver 
after unbind the particular pci device.

> ------------------------------------------------------------------------
> 
> In io/iommu/iommu_tests.py 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23discussion_r1006410229&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152905605%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=9bC8Mw%2BDHckikhR2TKgdrPX5CTexhEWcmq9U2vYI7Is%3D&reserved=0>:
> 
>> +        genio.write_file(f'/sys/bus/pci/devices/{pci_addr}/iommu_group/type',
> +                         dom)
> +        time.sleep(5)
> +        output = genio.read_one_line(
> +                f"/sys/bus/pci/devices/{pci_addr}/iommu_group/type")
> +        out = output.rsplit(None, 1)[-1]
> +        if (dom not in ("auto", out)) or (dom == "auto" and out != def_dom):
> +            self.fail(f'Domain type change failed {pci_addr}')
> +        else:
> +            self.log.info("successfully changed iommu group domain to %s", dom)
> +
> +    def bind(self, driver, pci_addr):
> +        """Bind the device to driver"""
> +        genio.write_file(f"/sys/bus/pci/drivers/{driver}/bind", pci_addr)
> +        time.sleep(5)
> +        os.listdir(f'/sys/bus/pci/drivers/{driver}')
> 
> same here.
> 

Similarily, by this line we are logging what are devices binded to 
driver after binding the particular pci device.

> ------------------------------------------------------------------------
> 
> In io/iommu/iommu_tests.py 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23discussion_r1006410410&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152905605%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=fIERcTgwKdP6ZAOCmOyEphzzfaA187HOQWCVhXA7xXY%3D&reserved=0>:
> 
>> +
> +    def change_domain(self, dom, def_dom, pci_addr):
> +        """Change the domain of device to dom"""
> +        genio.write_file(f'/sys/bus/pci/devices/{pci_addr}/iommu_group/type',
> +                         dom)
> +        time.sleep(5)
> +        output = genio.read_one_line(
> +                f"/sys/bus/pci/devices/{pci_addr}/iommu_group/type")
> +        out = output.rsplit(None, 1)[-1]
> +        if (dom not in ("auto", out)) or (dom == "auto" and out != def_dom):
> +            self.fail(f'Domain type change failed {pci_addr}')
> +        else:
> +            self.log.info("successfully changed iommu group domain to %s", dom)
> +
> +    def bind(self, driver, pci_addr):
> +        """Bind the device to driver"""
> 
> same here for docstring
> 
> ------------------------------------------------------------------------
> 
> In io/iommu/iommu_tests.py 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23discussion_r1006410810&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152905605%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=KF11YPw%2F33UyPBJdZU9Hq423TAE5aYiCfZY%2BUuqdWQY%3D&reserved=0>:
> 
>> +            self.fail(f'Domain type change failed {pci_addr}')
> +        else:
> +            self.log.info("successfully changed iommu group domain to %s", dom)
> +
> +    def bind(self, driver, pci_addr):
> +        """Bind the device to driver"""
> +        genio.write_file(f"/sys/bus/pci/drivers/{driver}/bind", pci_addr)
> +        time.sleep(5)
> +        os.listdir(f'/sys/bus/pci/drivers/{driver}')
> +        if not os.path.exists(f"/sys/bus/pci/drivers/{driver}/{pci_addr}"):
> +            self.fail(f'{pci_addr} not bounded back')
> +        else:
> +            self.log.info("successfully binded back %s", pci_addr)
> +
> +    def reset(self, pci_addr):
> +        """Remove the device"""
> 
> same docstring here
> 
> ------------------------------------------------------------------------
> 
> In io/iommu/iommu_tests.py 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23discussion_r1006411004&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152905605%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=JwzPlcPJctG1ojYdPzlH9RGnR4tdDu9rwXxqRYYJops%3D&reserved=0>:
> 
>> +        else:
> +            self.log.info("successfully binded back %s", pci_addr)
> +
> +    def reset(self, pci_addr):
> +        """Remove the device"""
> +        genio.write_file(f'/sys/bus/pci/devices/{pci_addr}/remove', '1')
> +        time.sleep(5)
> +        cmd = f"lspci -vvs {pci_addr}"
> +        output = process.run(cmd, ignore_status=True, shell=True).stdout_text
> +        if output != "":
> +            self.fail(f'Unsuccessful to remove {pci_addr}')
> +        else:
> +            self.log.info("successfully removed the device %s", pci_addr)
> +
> +    def rescan(self, pci_addr):
> +        """Rescan the system"""
> 
> same here, docstring
> 
> ------------------------------------------------------------------------
> 
> In io/iommu/iommu_tests.py 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23discussion_r1006411461&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152905605%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=5vqFpG6AC%2BobQZgS4qSi4sRtuJKzZMM2Q9WWsGBqdJA%3D&reserved=0>:
> 
>> +        else:
> +            self.log.info("successfully removed the device %s", pci_addr)
> +
> +    def rescan(self, pci_addr):
> +        """Rescan the system"""
> +        genio.write_file('/sys/bus/pci/rescan', '1')
> +        time.sleep(5)
> +        cmd = f"lspci -vvs {pci_addr}"
> +        output = process.run(cmd, ignore_status=True, shell=True).stdout_text
> +        if output == "":
> +            self.fail(f'Unsuccessful to rescan for {pci_addr}')
> +        else:
> +            self.log.info("successfully rescaned for the device %s", pci_addr)
> +
> +    def get_params(self, pci_addr):
> +        """Get device parameter-driver, group, default domain(def_dom)"""
> 
> docstring
> 
> ------------------------------------------------------------------------
> 
> In io/iommu/iommu_tests.py 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23discussion_r1006411625&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152905605%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=d6Qk4uKt28TzsU0%2FVJ5Wh%2FCIK4xFyLTHQjPP6ToMJLc%3D&reserved=0>:
> 
>> +
> +    def get_params(self, pci_addr):
> +        """Get device parameter-driver, group, default domain(def_dom)"""
> +        driver = pci.get_driver(pci_addr)
> +
> +        def_dom = None
> +        output = genio.read_one_line(
> +                f"/sys/bus/pci/drivers/{driver}/{pci_addr}/iommu_group/type")
> +        if output:
> +            def_dom = output.rsplit(None, 1)[-1]
> +        else:
> +            self.fail(f"Not able to get default domain of {pci_addr}")
> +        return driver, def_dom
> +
> +    def check(self, def_dom, pci_addr, driver):
> +        """Check if the PCI device is in default domain"""
> 
> same here
> 
> —
> Reply to this email directly, view it on GitHub 
> <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favocado-framework-tests%2Favocado-misc-tests%2Fpull%2F2434%23pullrequestreview-1157607670&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152905605%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pzI6O72%2BuYAA6sxQzyV8QitgpmgxhY107HW7Cr6uRsQ%3D&reserved=0>, or unsubscribe <https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FA3K6XNAFAO3JOX62MSUAN6TWFIGVHANCNFSM6AAAAAAQ3CLTH4&data=05%7C01%7Cdheerajkumar.srivastava%40amd.com%7C11712ad1a9354361012908dab7db63ba%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C638024450152905605%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=sBfCvS64X%2FVbpCtd1ndPXYKUdV1W1PAlC01n5Oub%2Beo%3D&reserved=0>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

Naresh-ibm commented:
@dhsrivas here is the snippet of this formate, please follow the same across the functions,

```
def create_raid(self, l_disk, l_raid_name):
        """
        creates a softwareraid with given raid name on given disk

        :param l_disk: disk name on which raid will be created
        :l_raid_name: name of the softwareraid
        :return: None
        """
```

dhsrivas commented:
Hi Naresh,
Based on your review comments, updated the commit. Pl. review.

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,336 @@
+                self.cancel("Device passed is not bound to any driver")
+            lst = os.listdir(f'/sys/bus/pci/drivers/{driver}/{pci_addr}/'
+                             'iommu_group/devices/')
+            if len(lst) != 1:
```
what if lenght is '0'  here?

dhsrivas replied:
This will never happen as we are looking in the iommu group of pci device passed, so pci device itself be there atleast in any case.

Naresh-ibm commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,336 @@
+                            "than one device but system doesnot support "
+                            "domain type change for such device")
+        cmd = "sudo dmesg -C"
+        process.run(cmd, ignore_status=True, shell=True)
```
add `'root=True'  `   instead of using sudo in cmd

dhsrivas replied:
Sure will update.

Naresh-ibm commented on io/iommu/iommu_tests.py.data/README.txt:
```diff
@@ -0,0 +1,19 @@
+Rescan - Rescan for pci device.
+
+Pci device to be tested (on subtests involving "change iommu group domain type") should be alone in its iommu group.
+Group domain type change for a pci device which belongs to an iommu group having more then one pci device is not supported
```
`more than/ more then` spelling

dhsrivas commented:
Addressed all the above comments - 

- Related to docstrings
- using wait_for() instead of sleep()
- self.log(os.listdir()) inplace of os.lisdir()

Pl. review the updated commit. Thanks.

Naresh-ibm commented:
@dhsrivas LGTM
please add the latest run logs for reference.

dhsrivas commented:
#sudo avocado run io/iommu/iommu_tests.py -p pci_devices="0000:03:00.3" -p coun
JOB ID     : 8faa1c5e818dac822d3d281fd996a508b86c90f2
JOB LOG    : /home/amd/workspace/IOMMU/IOMMU_Sanity/tests/results/job-2022-11-21T04.25-8faa1c5/job.log
 (1/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_bind: PASS (5.09 s)
 (2/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_changedomain_bind: PASS (60.17 s)
 (3/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_changedomain_ntimes_bind: PASS (180.40 s)
 (4/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_bind_rescan: PASS (5.08 s)
 (5/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_changedomain_bind_rescan: PASS (60.25 s)
 (6/6) io/iommu/iommu_tests.py:IommuTest.test_reset_rescan: PASS (0.07 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/amd/workspace/IOMMU/IOMMU_Sanity/tests/results/job-2022-11-21T04.25-8faa1c5/results.html
JOB TIME   : 311.24 s

Latest Log file - 
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/10053446/job.log)

PraveenPenguin commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,373 @@
+    if output != "":
+        return False
+    return True
+
```
why not push these in the avocado library itself , can be have in this utility 
https://github.com/avocado-framework/avocado/blob/master/avocado/utils/pci.py

dhsrivas replied:
Because of the issue https://github.com/avocado-framework/avocado/issues/5417, we are not able to use latest upstream avocado.
so, please merge this in tests now, we will take this as enhancement to avocado once that issue is fixed.

PraveenPenguin replied:
ah sure missed from my TODO, will have some thing on  YAML issue soon, but still, i say please have part of the utils itself

dhsrivas replied:
Yeah sure, once the issue is resolved will be moving the parts to avocado utils.

dhsrivas replied:
@PraveenPenguin Can u please the parts that we can have in avocado for reuse?

dhsrivas replied:
Seems, reset_check() and rescan_check() are two candidate to pci.py.
Let me know other functions to move to avocado/utils/pci.py.

PraveenPenguin commented on io/iommu/iommu_tests.py:
```diff
@@ -0,0 +1,373 @@
+                            "than one device but system doesnot support "
+                            "domain type change for such device")
+        cmd = "dmesg -C"
+        process.run(cmd, ignore_status=True, shell=True, sudo=True)
```
please dmesg library from avocado https://github.com/avocado-framework/avocado/blob/master/avocado/utils/dmesg.py

dhsrivas replied:
>

Avocado function to clear dmesg uses "dmesg -c" command which will first show the dmesg then clears it. So for two reason not included that
1) This will show all the system dmesgs that are there before the start of first subtest case.
2) Do we require all the dmesgs for each subtest to be printed. As of now we are mainly concerned with dmesgs with levels=warning, critical, error and alert. These if present are printed and test fails.

So we used instead "dmesg -C" which simply clears the dmesg. Thats the reason of not using inbuilt function in avocado.

PraveenPenguin replied:
sure in that case we should be enhancing the library may have function argument  in the library which can decide this scenario , please let me know you need more info on this as i love to see this should be implement in library

dhsrivas replied:
Sure.

PraveenPenguin reviewed (commented):
@dhsrivas sorry for the late review, I see places that can be moved to avocado, as that will help us to reuse,

PraveenPenguin commented:
@dhsrivas as i pointed some places i see more instances that should also be ideal candidate for same

dhsrivas commented:
@Naresh-ibm If everything is fine. Pl. merge.

Naresh-ibm commented:
@PraveenPenguin @narasimhan-v Please ACK if no more comments from your side.
will merge it.

narasimhan-v commented:
> @PraveenPenguin @narasimhan-v Please ACK if no more comments from your side.
> will merge it.

ACK

PraveenPenguin commented:
> @PraveenPenguin @narasimhan-v Please ACK if no more comments from your side. will merge it.

I am still interested roadmap around putting some of the code as part of avocado utility, if not now we should have TODO as a comment in test, and later we can pick that up

dhsrivas commented:
@Naresh-ibm If everything is fine, pl. approve and merge.

Naresh-ibm reviewed (approved):
@PraveenPenguin @narasimhan-v Thanks for all suggestion and reviews, and @dhsrivas for accommodating all changes patiently. 

LGTM
